lighthouse-core/audits/legacy-javascript.js

```diff
@@ -0,0 +1,334 @@
+      'String.prototype.codePointAt',
+      'String.prototype.endsWith',
+      'String.fromCodePoint',
+      'String.prototype.includes',
+      'String.raw',
+      'String.prototype.repeat',
+      'String.prototype.startsWith',
+      'String.prototype.trim',
+      'ArrayBuffer',
+      'DataView',
+      'Float32Array',
+      'Float64Array',
+      'Int16Array',
+      'Int32Array',
+      'Int8Array',
+      'Uint16Array',
+      'Uint32Array',
+      'Uint8Array',
+      'Uint8ClampedArray',
+      'WeakMap',
+      'WeakSet',
+      'Array.prototype.includes',
+      'Object.entries',
+      'Object.getOwnPropertyDescriptors',
+      'Object.values',
+      'String.prototype.padEnd',
+      'String.prototype.padStart',
+    ].map(str => {
+      const parts = str.split('.');
+      const object = parts.length > 1 ? parts.slice(0, parts.length - 1).join('.') : null;
+      const property = parts[parts.length - 1];
+      return {
+        name: `${object ? object + '.' : ''}${property}`,
+        expression: this.buildPolyfillExpression(object, property),
+      };
+    });
+  }
+
+  /**
+   * @return {Pattern[]}
+   */
+  static getTransformPatterns() {
+    return [
+      {
+        name: '@babel/plugin-transform-classes',
+        expression: 'Cannot call a class as a function',
+      },
+      {
+        name: '@babel/plugin-transform-regenerator',
+        expression: 'regeneratorRuntime.wrap',
+      },
+      {
+        name: '@babel/plugin-transform-spread',
+        expression: /\.apply\(void 0/.source,
+      },
+    ];
+  }
+
+  /**
+   * @param {CodePatternMatcher} matcher
+   * @param {LH.GathererArtifacts['ScriptElements']} scripts
+   * @param {LH.Artifacts.NetworkRequest[]} networkRecords
+   * @return {{
+   *  patternCounter: Map<string, number>,
+   *  patternMatchCounter: Map<PatternMatchResult, number>,
+   *  urlToMatchResults: Map<string, PatternMatchResult[]>,
+   * }}
+   */
+  static detectCodePatternsAcrossScripts(matcher, scripts, networkRecords) {
+    /** @type {Map<string, number>} */
+    const patternCounter = new Map();
+    /** @type {Map<PatternMatchResult, number>} */
+    const patternMatchCounter = new Map();
+    /** @type {Map<string, PatternMatchResult[]>} */
+    const urlToMatchResults = new Map();
+
+    for (const {requestId, content} of Object.values(scripts)) {
+      if (!content) continue;
+      const networkRecord = networkRecords.find(record => record.requestId === requestId);
+      if (!networkRecord) continue;
+      const matches = matcher.match(content);
+      if (!matches.length) continue;
+
+      urlToMatchResults.set(networkRecord.url, matches);
+      for (const match of matches) {
+        const val = patternCounter.get(match.name) || 0;
+        patternMatchCounter.set(match, val);
+        patternCounter.set(match.name, val + 1);
+      }
+    }
+
+    return {patternCounter, patternMatchCounter, urlToMatchResults};
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
```

Comment:
Definite need a comment here, this `audit` is very heavy and needs some explanation, maybe even an _ascii diagram_!